lib/Pass/Pass.cpp

```diff
@@ -133,27 +150,32 @@ struct SouperPass : public FunctionPass {
 
     appendToGlobalCtors(*M, Ctor, 0);
 
+    BasicBlock::iterator BI = I;
+    while (isa<PHINode>(*BI))
+      ++BI;
     new AtomicRMWInst(AtomicRMWInst::Add, CntVar,
                       ConstantInt::get(C, APInt(64, 1)), Monotonic, CrossThread,
                       BI);
   }
 
-  bool runOnFunction(Function &F) {
-    bool changed = false;
+  std::set <Function *> DoneFuncs;
+
+  bool runOnFunction(Function *F) {
+    bool Changed = false;
     InstContext IC;
     ExprBuilderContext EBC;
     CandidateMap CandMap;
-
-    LoopInfo *LI = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
-
-    FunctionCandidateSet CS = ExtractCandidatesFromPass(&F, LI, IC, EBC);
+    LoopInfo *LI = &getAnalysis<LoopInfoWrapperPass>(*F).getLoopInfo();
+    if (!LI)
+      report_fatal_error("getLoopInfo() failed");
+    FunctionCandidateSet CS = ExtractCandidatesFromPass(F, LI, IC, EBC);
 
     std::string FunctionName;
-    if (F.hasLocalLinkage()) {
+    if (F->hasLocalLinkage()) {
       FunctionName =
-        (F.getParent()->getModuleIdentifier() + ":" + F.getName()).str();
+        (F->getParent()->getModuleIdentifier() + ":" + F->getName()).str();
     } else {
-      FunctionName = F.getName();
+      FunctionName = F->getName();
     }
```

Comment:
Overall I don't think we're very consistent about this choice. We can address this later with a cleanup patch (though I don't really have strong feelings about doing this) but I think for now this is fine. Thanks for the careful read!